Use msgp for mpt value node serialization

peterlimg opened the pull request:
# Fixes: 

- #1003 
- Fix [newRead](https://github.com/0chain/0chain/blob/15b1511a03e10fa38c8a55fd4a12ab71e627f2b4/code/go/0chain.net/smartcontract/storagesc/stats.go#L27) and [newChallenge](https://github.com/0chain/0chain/blob/15b1511a03e10fa38c8a55fd4a12ab71e627f2b4/code/go/0chain.net/smartcontract/storagesc/stats.go#L46) issues, they should create a new node and save to MPT when get 'value not present' error instead of return error, and what's worse was that the errors were not checked. So the caller thought that the API was executed successfully, while actually it didn't.

# Changes:
- Changed the MPT value node serialization package from json to msgp (https://github.com/tinylib/msgp), we are using a  forked version on https://github.com/0chain/msgp, with the support of `time.Duration` and `sorted map`.  
- Added a new `util.MPTSerializable` interface, which will be used by MPT only to do the serialization. e 
- When we need to add a new data struct to MPT, for example:

    ```go
    //go:generate msgp -io=false -tests=false -v
    type Foo struct {
        Name string
    }
    ```
         
    we don't have to write the `Encode` and `Decode` methods manually like we do before, instead we can add a `//go:generate msgp -io=false -tests=false -v` in the file where the struct is defined. Then run the following command from the project root to automatically generate methods for serialization. 
    
    ```bash
   make msgp
    ```
    
    A new file will then be generated as `{file}_gen.go` in the same dir.



- Changed the `GetTrieNode(key datastore.Key) (util.Serialization, error)` interface to `GetTrieNode(key datastore.Key, v util.MPTSerializable) error`. So the node value will be returned by the second param, i.e `v util.MPTSerializable` just like what `json.Unmarshal()` do, in this way, we don't have to get data bytes from MPT first, then decode it manually everywhere. 
- There is one thing to note when using the msgp, 
  - It does not support the type alias for system types, therefore the `type Key = string` in `datastore` package is not supported, I changed the type of `datastore.Key` to string. 
 
---
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master BEFORE merging this PR  
  
Associated PRs (Link as appropriate):
...

peterlimg committed:
Replace json with msgp(msgpack) for MPT value data serialization

peterlimg committed:
Merge staging

peterlimg committed:
Fix wrong mockery input params

peterlimg committed:
Commit missing msgp *_gen.go files

peterlimg committed:
Use system temp dir for testing

peterlimg committed:
Fix errors

peterlimg committed:
Pass linter, all errors will be fixed in a new PR

peterlimg committed:
Do not use omitempty for mspg

peterlimg committed:
Add a CI script to check if there are msgp changes

Sriep commented on code/go/0chain.net/chaincore/chain/state/state_context.go:
```diff
@@ -264,15 +265,15 @@ func (sc *StateContext) Validate() error {
 	s := &state.State{}
-	s.Balance = state.Balance(0)
-	ss, err := sc.state.GetNodeValue(util.Path(clientID))
+	//s.Balance = state.Balance(0)
```
Remove commented-out code before merging.

Sriep commented on code/go/0chain.net/chaincore/chain/entity.go:
```diff
@@ -97,14 +97,14 @@ type Chain struct {
 	BlocksToSharder int
 
-	MagicBlockStorage round.RoundStorage `json:"-"`
+	MagicBlockStorage round.RoundStorage `json:"-" msg:"-"`
```
Why is there no  `//go:generate msgp -io=false -tests=false -v` in this file?

peterlimg replied:
The Chain struct is not saved in MPT, this tag was added in case it needs to be serialized with msgp in the future. Perhaps I should remove the tag for now.

Sriep replied:
The important thing is to make sure `//go:generate msgp` appears everywhere it should. Other than confusing me, extra `msg` tags do not seem to pose a problem.

Sriep commented on code/go/0chain.net/chaincore/chain/state_handler.go:
```diff
@@ -74,30 +71,32 @@ func (c *Chain) GetSCRestOutput(ctx context.Context, r *http.Request) (interface
 }
 
 func (c *Chain) GetNodeFromSCState(ctx context.Context, r *http.Request) (interface{}, error) {
```
What is the story here? Maybe replace the commented out code with a not implemented any more error.

It seems to be a handler for `"/v1/scstate/get"`. Is this API call no longer supported or is it a todo?

peterlimg replied:
This was commented out to pass compiler and do tests. Anyway, fixed in new commits.

Sriep commented on code/go/0chain.net/chaincore/state/partial_state_test.go:
```diff
@@ -631,36 +629,36 @@ func newValueNode(version string, v int64) *valueNode {
-		require.Equal(t, int64(2022), vv.Value)
-	}
-}
+//func TestPartialUnmarshalMsgpack(t *testing.T) {
```
Remove commented out code.

Sriep commented on code/go/0chain.net/chaincore/transaction/entity.go:
```diff
@@ -46,14 +46,14 @@ func SetupTransactionDB(redisTxnsHost string, redisTxnsPort int) {
 type Transaction struct {
 	datastore.HashIDField
-	datastore.CollectionMemberField `json:"-" msgpack:"-"`
+	datastore.CollectionMemberField `json:"-" msgpack:"-" msg:"-"`
```
`//go:generate msgp`? Why don't we want to generate this file, when we are adding `msg` tags?

peterlimg replied:
This struct is not saved on MPT, I will remove the tag, and add them if we needed in the future.

Sriep commented on code/go/0chain.net/core/util/merkle_patricia_trie_test.go:
```diff
@@ -546,38 +543,39 @@ func TestMPT_MultipleConcurrentInserts(t *testing.T) {
 }
 
 func TestMPT_ConcurrentMerges(t *testing.T) {
```
If this test no longer makes sense we should remove it.

Sriep commented on code/go/0chain.net/core/util/merkle_patricia_trie_test.go:
```diff
@@ -970,7 +980,7 @@ func TestAddTwiceDeleteOnce(t *testing.T) {
 
 func TestWithPruneStats(t *testing.T) {
-	t.Parallel()
+	////t.Parallel()
```
we can remove these comments.

Sriep commented on code/go/0chain.net/core/util/mpt_node.go:
```diff
@@ -274,27 +269,54 @@ func (ln *LeafNode) Decode(buf []byte) error {
-	return ln.Value != nil && ln.Value.HasValue()
+	// TODO: remove the GetValueBytes check, as it would try to marshal the value to see if it is empty, which is
+	// a waste of time
+	//return ln.Value != nil && len(ln.Value.GetValueBytes()) > 0
```
Can't you remove the GetValueBytes comment now, and satisfy the todo?

Sriep commented on code/go/0chain.net/core/util/mpt_nodedb_test.go:
```diff
@@ -494,159 +494,161 @@ func TestLevelNodeDB_Current_Prev_Rebase(t *testing.T) {
-	})
-
-}
+//func TestPNodeDB_Full(t *testing.T) {
```
Remove test if we don't need it any more.

Sriep replied:
can remove comment.

Sriep commented on code/go/0chain.net/core/util/mpt_nodedb_test.go:
```diff
@@ -696,65 +698,75 @@ func noNodeNotFound(err error) error {
 }
 
 // launch test with -race or make the case sense
```
Remove test if we don't need it any more.

Sriep commented on code/go/0chain.net/core/util/mpt_pnodedb_test.go:
```diff
@@ -12,47 +11,52 @@ import (
+//	return nil
+//}
 
 func TestNewPNodeDB(t *testing.T) {
```
Remove test if we don't need it any more.

Sriep commented on code/go/0chain.net/sharder/block_test.go:
```diff
@@ -44,15 +44,15 @@ const (
-	require.NoError(t, err)
-
-	dbDir, err := ioutil.TempDir(tmpDir, "dbs")
+	//cd, err := os.Getwd()
```
remove commented code. Also end of function.

Sriep replied:
can remove comment

Sriep commented on code/go/0chain.net/smartcontract/zcnsc/context_helpers_test.go:
```diff
@@ -93,45 +195,41 @@ func MakeMockStateContext() *mocks.StateContextI {
-			func(_ datastore.Key) error {
-				return nil
-			})
+	//for _, an := range authorizers {
```
Remove code if its not needed.

Sriep commented on code/go/0chain.net/smartcontract/zcnsc/models.go:
```diff
@@ -445,10 +301,19 @@ func (un *UserNode) Save(balances cstate.StateContextI) (err error) {
 
 // ----------  TokenLock -----------------
 
+//type ZcnLockingPool struct {
```
comment not needed.

Sriep reviewed (commented):
I notice that often we add `msg` tags to objects but then don't set to generate a `_gen.go` file with the `//go:generate msgp ` comment.

Is this an oversight or are `msg` tags used other than the `_gen.go` file generater?

peterlimg committed:
Remove unnecessary msgp:"-" tag

peterlimg committed:
Add mpt.GetNodeValueRaw to return raw data bytes slice

to fix the /v1/scstate/get handler.

This api would return any state node value in json format, however, msgp does not support decoding []byte into an raw interface{} object. The workaround is calling msgp.UnmarshalAsJSON() to conver the msgpack data bytes into json format, then use json decoder to unmarshal to the raw interface as it did before.

Sriep commented on code/go/0chain.net/chaincore/chain/protocol_view_change.go:
```diff
@@ -152,15 +152,15 @@ func (mc *Chain) isRegisteredEx(ctx context.Context, getStatePath func(n *node.N
-				zap.Any("error", err))
-			return false
-		}
+		//if list == nil {
```
remove comment

peterlimg replied:
> It might be an idea to cearafully merge in recent staging changes. I saw one possable issue with [storagesc/stakepool.go](https://github.com/0chain/0chain/pull/1033/files?show-deleted-files=true&show-viewed-files=true&file-filters%5B%5D=#diff-ce2ed4fb820ddf8bc8517075d5242e73b2e7547262aababb0be2e39ea481a56dR28) there could be others.

Yes, basically could happen if I did some changes here, and then this code was removed from staging. Don't know why the git didn't detect the conflicts. Anyway, it should be safe as this code is unused, I'll fix them.

Sriep commented on code/go/0chain.net/chaincore/chain/protocol_view_change.go:
```diff
@@ -602,8 +602,8 @@ func (c *Chain) GetPhaseFromSharders(ctx context.Context) {
 
-	var seri util.Serializable
-	seri, err = c.GetBlockStateNode(b, minersc.PhaseKey)
+	//var seri util.Serializable
```
remove comment

Sriep commented on code/go/0chain.net/chaincore/chain/state_handler_test.go:
```diff
@@ -59,6 +57,61 @@ func init() {
+		InterestInterval time.Duration `json:"interest_interval"`
+	}
+
+	//scConfig struct {
```
can remove comment

Sriep commented on code/go/0chain.net/chaincore/state/signed_transfer.go:
```diff
@@ -68,7 +68,7 @@ func (st SignedTransfer) verifyPublicKey() error {
 	}
 
-	if encryption.Hash(publicKeyBytes) != st.Transfer.ClientID {
+	if encryption.Hash(publicKeyBytes) != string(st.Transfer.ClientID) {
```
string typecast does not seem necessary.

peterlimg replied:
yes, don't know why I did this, lol

Sriep commented on code/go/0chain.net/chaincore/wallet/wallet_test.go:
```diff
@@ -244,23 +243,24 @@ func createWallets(num int) []*Wallet {
-		}
-		return s, nil
-	}
+	//s, ok := ss.(*state.State)
```
can remove comment

Sriep commented on code/go/0chain.net/core/common/time_gen_test.go:
```diff
@@ -0,0 +1,3 @@
+package common
+
+// Code generated by github.com/tinylib/msgp DO NOT EDIT.
```
This file appears empty, do we need it?

Sriep commented on code/go/0chain.net/core/util/merkle_patricia_trie_test.go:
```diff
@@ -201,15 +203,10 @@ func doStateValInsert(t *testing.T, mpt MerklePatriciaTrieI, key string, value i
+	astate := &AState{}
+	err := mpt.GetNodeValue([]byte(key), astate)
+	assert.NoError(t, err)
+	//assert.NoError(t, astate.Decode(val.Encode()))
```
can remove comment

Sriep commented on code/go/0chain.net/core/util/mpt_node_test.go:
```diff
@@ -1,422 +1 @@
 package util
```
file appears empty, we should probably remove it.

Sriep commented on code/go/0chain.net/smartcontract/storagesc/stakepool.go:
```diff
@@ -21,6 +22,108 @@ import (
+//go:generate msgp -io=false -tests=false -unexported=true -v
+
+// A UserStakePools collects stake pools references for a user.
+type UserStakePools struct {
```
Not sure we need this code, it could be a merging issue? Check https://github.com/0chain/0chain/pull/967.

peterlimg replied:
Yes, not needed, I'll remove it.

Sriep commented on code/go/0chain.net/smartcontract/zcnsc/authorizer_test.go:
```diff
@@ -350,12 +347,14 @@ func Test_NewAuthorizer_MustHave_LockPool_Initialized(t *testing.T) {
+	err = ctx.GetTrieNode(node.GetKey(), newNode)
 	require.NoError(t, err)
+
+	//require.NotNil(t, blob)
```
can remove

Sriep commented on code/go/0chain.net/smartcontract/zcnsc/mint.go:
```diff
@@ -53,31 +54,54 @@ func (zcn *ZCNSmartContract) Mint(trans *transaction.Transaction, inputData []by
 		err = common.NewError(code, fmt.Sprintf("get user node error (%v), %s", err, info))
 		return
 	}
+	//if err != nil && payload.Nonce != 1 {
```
remove commented out code

Sriep commented on code/go/0chain.net/smartcontract/zcnsc/nodes_reader.go:
```diff
@@ -24,74 +22,65 @@ func isNil(i interface{}) bool {
 	}
-	return node, nil
+
+	//if err != nil {
```
can remove

Sriep commented on code/go/0chain.net/chaincore/chain/state/state_context.go:
```diff
@@ -264,15 +265,14 @@ func (sc *StateContext) Validate() error {
 		return s, err
 	}
-	s = sc.clientStateDeserializer.Deserialize(ss).(*state.State)
+	//s = sc.clientStateDeserializer.Deserialize(ss).(*state.State)
```
can remove

Sriep reviewed (commented):
It might be an idea to carefully merge in recent staging changes. I saw one possible merging issue with [storagesc/stakepool.go](https://github.com/0chain/0chain/pull/1033/files?show-deleted-files=true&show-viewed-files=true&file-filters%5B%5D=#diff-ce2ed4fb820ddf8bc8517075d5242e73b2e7547262aababb0be2e39ea481a56dR28) there could be others.

peterlimg committed:
Remove unused code in storagesc/stakepool.go

peterlimg committed:
Remove unused commented code

peterlimg committed:
Clean msgp:"-" tag

peterlimg committed:
Clean up commented code

peterlimg committed:
Revert unit test fix, need further refactoring on

Handlers to fix the test issue

service-0chain commented:
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | feature/use-msgp-for-mpt  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

peterlimg committed:
Fix faucet sc 'pour' error

Empty userNode was not checked which caused the miner panic.

peterlimg committed:
Change partition list back to private

It was changed to public before I knew msgp could support unexported types serialization by using `-unexported=true` option.

peterlimg committed:
Fix value not present on getting global node of faucetsc

peterlimg committed:
Fix newRead/newWrite/newChallenge

Insert new to MPT if err is ErrValueNotPresent

peterlimg committed:
Fix node pool decoding that could cause state hash mismatch error

peterlimg committed:
Fix random list decode issue

peterlimg committed:
Fix UserNode json decode issue that could

cause hash state mismatch error

peterlimg committed:
Remove unnecessary msgp -unexported flag

peterlimg committed:
Use msgp from peterlimg/msgp with sorted map encoding

The tinylib/msgp encode map without sorting, which caused the 'state hash mismatch' error

peterlimg committed:
Update msgp gen code for new 'cost' field

peterlimg committed:
Fix panic in ComputeFinalizedBlock

peterlimg committed:
Fix inter_pool decode error

peterlimg committed:
Fix conflicts after merging staging

peterlimg committed:
Fix unit test errors

peterlimg committed:
Add events on block finalizing process

peterlimg committed:
Change back to use time.Duration

Update msgp to support time.Duration

peterlimg committed:
Fix conflicts after merging staging

peterlimg committed:
Remove debug code

peterlimg committed:
Fix unit tests

peterlimg committed:
Pass linter

service-0chain commented:
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | fix/verify-miner-fees-test  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | feature/use-msgp-for-mpt  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

peterlimg committed:
Change to use 0chain/msgp

Sriep commented on code/go/0chain.net/core/datastore/collection_entity.go:
```diff
@@ -67,7 +67,7 @@ func (d *EntityCollection) GetCollectionName(parent Key) string {
 	name := d.CollectionName
 	if !IsEmpty(parent) {
-		name += ":" + parent
+		name += ":" + string(parent)
```
It might be better to change `parent` to string instead of the typecast
```go
func (d *EntityCollection) GetCollectionName(parent string) string {
```
The only two times GetCollectionName is called it passes [""](https://github.com/0chain/0chain/blob/feature/use-msgp-for-mpt/code/go/0chain.net/core/datastore/collection_member_field.go#L21) and [t.ChainID](https://github.com/0chain/0chain/blob/feature/use-msgp-for-mpt/code/go/0chain.net/chaincore/transaction/entity.go#L222) which is a [string](https://github.com/0chain/0chain/blob/feature/use-msgp-for-mpt/code/go/0chain.net/chaincore/transaction/entity.go#L56).

Sriep commented on code/go/0chain.net/core/util/merkle_patricia_trie_test.go:
```diff
@@ -1058,7 +1030,7 @@ func TestCloneMPT(t *testing.T) {
 		tt := tt
 		t.Run(tt.name, func(t *testing.T) {
-			t.Parallel()
+			//////t.Parallel()
```
Maybe we should clean up these `///////t.Parallel()` comments somehow.

Sriep commented on code/go/0chain.net/smartcontract/storagesc/expose_mpt.go:
```diff
@@ -1,33 +1,34 @@
-			"get trie node %s failed: %v", key, err)
-	}
-	return string(val.Encode()), nil
+	//var err error
```
This commented out code does not work anymore. I see no reason to keep it.

Sriep commented on code/go/0chain.net/smartcontract/storagesc/stakepool.go:
```diff
@@ -169,10 +157,10 @@ func (sp *stakePool) extendOffer(delta state.Balance) (err error) {
+//type stakePoolUpdateInfo struct {
+//	stake  state.Balance // stake of all delegate pools
+//	offers state.Balance // offers stake
+//}
```
Leave untouched or remove code; make future merging easier.

Sriep commented on code/go/0chain.net/smartcontract/vestingsc/vesting_test.go:
```diff
@@ -119,8 +146,8 @@ func TestVestingSmartContract_getPoolBytes_getPool(t *testing.T) {
-	_, err = vsc.getPoolBytes(poolKey(vsc.ID, txHash), balances)
-	require.Equal(t, util.ErrValueNotPresent, err)
+	//_, err = vsc.getPoolBytes(poolKey(vsc.ID, txHash), balances)
+	//require.Equal(t, util.ErrValueNotPresent, err)
```
can remove.

Sriep reviewed (commented):
Should be merged after https://github.com/0chain/0chain/pull/1059

peterlimg committed:
Clean up for code review

peterlimg commented:
Since there are a few new data struct created in #963 , I'll hold on this PR till the 963 is merged and resolve conflicts on top of it. @Sriep thanks for you review, I'll ping you later after #963 is done.

peterlimg committed:
Add make commands to install and run msgp

peterlimg committed:
Merge branch 'staging' into feature/use-msgp-for-mpt
